packages/transformer-directives/src/apply.ts

```diff
@@ -0,0 +1,98 @@
+import type { StringifiedUtil, UnoGenerator } from '@unocss/core'
+import { expandVariantGroup, notNull, regexScopePlaceholder } from '@unocss/core'
+import type { CssNode, Rule, Selector, SelectorList } from 'css-tree'
+import type MagicString from 'magic-string'
+import { clone, generate, parse } from 'css-tree'
+import type { TransformerDirectivesOptions } from '.'
+
+type Writeable<T> = { -readonly [P in keyof T]: T[P] }
+
+interface ApplyContext {
+  offset?: number
+  code: MagicString
+  uno: UnoGenerator
+  childNode: CssNode
+  node: Rule
+}
```

Comment:
I think we can combine these three contexts into one context. and in `ApplyContext` just put other params into function arguments.